src/playground-service-worker-proxy.ts

```diff
@@ -52,73 +52,122 @@ import {
     {scope}
   );
 
-  window.addEventListener(
-    'message',
-    (event: MessageEvent<PlaygroundMessage>) => {
-      if (event.data.type === UPDATE_SERVICE_WORKER) {
-        // When the project handshakes with the service worker, it may notice a
-        // version mismatch, in which case it will send this message to request
-        // the service worker update.
-        //
-        // Calling update triggers a fetch of the service worker. If the scripts
-        // are byte-different, then it will be installed + activated, and we'll
-        // receive the "updatefound" even there. Note this is exactly the normal
-        // Service Worker update lifecycle, it's just that calling update()
-        // forces it to happen immediately, as opposed to at some unknown
-        // opportunistic time in the future.
-        //
-        // TODO(aomarks) It would be good to show an error to the user somehow
-        // if we never end up getting a new version here. That would indicate
-        // that the SW and project are out of date *on the server*.
-        registration.update();
-        // There seems to be a bug in Chrome where the previous version of a
-        // service worker sometimes gets stuck and never shuts down when an
-        // updating is pending, even across a full browser restart. Calling
-        // unregister here seems to reliably force the previous version to shut
-        // down.
-        registration.unregister();
-      }
+  /** https://www.w3.org/TR/service-workers/#get-newest-worker-algorithm */
+  const getNewestWorker = () =>
+    registration.installing ?? registration.waiting ?? registration.active;
+
+  /** The service worker we most recently connected to. */
+  let newest: ServiceWorker | null = null;
+
+  /**
+   * Resolve when the given service worker reaches its activated state.
+   */
+  const activated = async (sw: ServiceWorker): Promise<void> => {
+    if (sw.state === 'activated') {
+      return;
     }
-  );
+    return new Promise<void>((resolve) => {
+      const aborter = new AbortController();
+      sw.addEventListener(
+        'statechange',
+        () => {
+          if (sw.state === 'activated') {
+            resolve();
+            aborter.abort();
+          }
+        },
+        {signal: aborter.signal}
+      );
+    });
+  };
 
-  const connect = (sw: ServiceWorker) => {
-    const {port1, port2} = new MessageChannel();
+  const connectToNewest = async () => {
+    const sw = getNewestWorker();
+    newest = sw;
```

Comment:
Good question! I originally factored it the way you suggested, but we need to account for the `UPDATE_SERVICE_WORKER` case at the bottom of this file. In that case, we need to force a reconnection even though we were previously connected (for when the service worker goes to sleep, then wakes up with amnesia).

Since now I think it actually is kinda unclear, I refactored it as described. There's now a force parameter which hopefully makes this a little more self describing than before.

I also renamed `newest` to `connected` because I think that makes a little more sense. And moved a variable down a little.